### PR TITLE
[master < T0945DX] Add Neo4j client and mixed workload to mgbench

### PR DESCRIPTION
**Pull request by antejavor**
In this PR, the following things has been done: 

- Added support for running Neoj4 (this introduces `--vendor-name` and change  from `--memgraph-binary` to `--vendor-binary`
- Flag for support when does authorization runs with 
- Support for and flag for warmup run
- Support and flag for mixed workload, run mixed queries from different categories. 
- Support and flag for tail latency 
- Support and flag for performance tracking (logs information about RES, through query/workload execution)
- `compare_results` support for mixed tail latency
- Dataset setup (index + delete) and import are separated. 
- Added `graph_bench.py` for executing multiple workloads on single vendor, sample:
```
graph_bench.py
--vendor memgraph /home/memgraph/binary
--dataset-group basic
--dataset-size small
--realistic 100 30 70 0 0
--realistic 100 50 50 0 0
--realistic 100 70 30 0 0
--realistic 100 30 40 10 20
--mixed 100 30 0 0 0 70
```

### CODE REVIEW

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:222]: Put this under if or some other nicer concept, since you have 2 vendors, if seems quite ok from my perspective 😄

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:228]: 0 commented code please 😄

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:262]: This is very strange (from before), that hole block could/should be massively simplified

File: tests/mgbench/runners.py
Comment (gitbuda) [inline tests/mgbench/runners.py:147]: make the code work or remove the commented code completly

**Review by gitbuda — CHANGES_REQUESTED** [review]
I've put a few comments +

@42jeremy it seems to me that the best place for queries that are different for different vendors is the `Dataset` class, the derived one because all the queries are in the same spot and you can see the different immediately, "centralization" makes total sense in this case?

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: I updated this and now changed is with arguments.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: When setting trigger on neo4j, if it is there you get an exepction, I uncommented this, but I will updated dataset with this line.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:303]: Yes, I have an idea how to make it a bit different, I will change tomorrow/or next week.

File: tests/mgbench/datasets.py
Comment (gitbuda) [inline tests/mgbench/datasets.py:489]: two updates at the end, one is enough

File: tests/mgbench/datasets.py
Comment (gitbuda) [inline tests/mgbench/datasets.py:483]: one write is fine as well 😄 (maybe there are more queries like that)

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:270]: query in not used in the loop, so use `_`

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:267]: `analytical, read, update, write = [], [], [], []`

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:290]: Ideas:
* I would write a function for a single check (`len(write) == 0 and distribution != 0`) and then use https://docs.python.org/3/library/functions.html#any
* This whole block of code (list init + for + this if) seems very repetitive and error-prone + since there are `ops`, `distributions`, maybe there is a nice/short abstraction for that? But that might be too much for this simple piece of logic here...

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:299]: There has to be a better and more clean way 😄

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:303]: Aha, `log` is used here! Be consistent with print VS log 😄

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:370]: The flag vs env var for authorization?

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:550]: For the mixed_workload types, I think it would be better to use an enum with specific values, which then can be nicely named and checked in these ifs...

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:714]: Flag?

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:423]: Flag?

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:753]: Feel free to implement 😄

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:745]: Authorization flag

File: tests/mgbench/benchmark.py
Comment (gitbuda) [inline tests/mgbench/benchmark.py:352]: Also repetitive (and also the `if ... continue` block above), maybe there could be some sort of an abstract definition, e.g., the `Op` (read, write, ...) and `Workload` (a collection of queries), that would simplify/abstract all the logic about generating the queries. NOTE: if you go there, apply the change to all query generation code.

**Review by gitbuda — CHANGES_REQUESTED** [review]
Overall not bad, a few things for sure can be improved, not all my comments are required except the ones that make the code more correct and consistent, but ofc let's comment one by one 😄

File: tests/mgbench/benchmark.py
Comment (42jeremy) [inline tests/mgbench/benchmark.py:142]:
```suggestion
    selected from the appropriate groups 
```

File: tests/mgbench/benchmark.py
Comment (42jeremy) [inline tests/mgbench/benchmark.py:566]:
```suggestion
    # TODO: Create some abstract class for vendors, that will hold this data
```

File: tests/mgbench/datasets.py
Comment (42jeremy) [inline tests/mgbench/datasets.py:50]: I guess the formatting is automatic. I think we all need to use the same configuration for that, my automatic formatting seems to give another layout. I'll look at it

**Comment by 42jeremy** [comment]
I have reviewed and haven't found big issues. However I think I'm missing the big picture, perhaps we should jump on a call and discuss the big lines so I can perform a higher level review.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:32]: Regarding this flag, currently on the master running setup is not configurable you were running auth code and non auth code on each run.  Now, you can choose not to run auth code if you do not want. Probably running both should be default?  These strings `"with_fine_grained_authorization"` and `"without_fine_grained_authorization"` are needed because of the graphana and I need them in JSON structure. What I can change  to drop global vars or local vars is is to change a flag to something like this: 
```suggestion
parser.add_argument(
    "--authorization",
    choices=["with_fine_grained_authorization", "without_fine_grained_authorization"],
    default="without_fine_grained_authorization",
    help="Run each query with authorization",
)
```
But this means running just auth or not running auth, not both.

From my perspective, the real questions are:
1. Do we want to run auth as default? Now it pops to me that we should because of graphana. 
2.  Do we need tail latency on auth? Currently, we are not running it.  I think we probably should.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: The only change here is that you do not need to specify the query, because it will be automatically executed on every query from group. 
```
./benchmark.py --mixed-workload=[1000, 10, 10, 30, 50, 0] pokec/small/group/**query** 

./benchmark.py --mixed-workload=[1000, 30, 0, 0, 0, 70] pokec/small/group/**query**
```

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:152]: I agree with you on this one 100%,  I was thinking about this at some point, but I prioritize other stuff for getting the results.  After merging this, I will open a separate PR for the logger, so we can log some extra stuff for the things you mention.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:221]: Will do in the next commit.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:303]: It is not the log you think it is. Take a look at what is log here: https://github.com/memgraph/memgraph/blob/master/tests/mgbench/log.py#L21
I will implement the right one in next PR.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: Ok, I fixed this by introducing simple Workload class, that now handles this info about number of arguments: https://github.com/memgraph/memgraph/pull/566/commits/d075b31040b6f67fa1d248f664b48899bed1baeb#diff-7bdbec771008d0634e8a5ecf53585bd3182efe304cf89a2809c1172d78ff9237R158

This defines the type of workload by name, so it is more readable.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: Regarding the cleaning of repetitive code, I cleaned it up by using dictionaries.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: I used the Workload class for that now.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: I am using a Class for this since Enum is limited in construction logic. Regarding the naming, on the graph bench we are using names: 

- Isolated  - all pure queries 
- Mixed - all queries + some percentage of  most likely write 
- Realistic - groups of queries in different combinations write 30% read 30% update 10% etc. 

I am pointing this out, bcs we have single -mixed-workload flag for two similar but different workloads. We are calling different names on the graph bench platform.  Maybe we can separate this, but obviously not super important.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:689]: https://github.com/memgraph/memgraph/pull/566#discussion_r1016135559

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:405]: https://github.com/memgraph/memgraph/pull/566#discussion_r1016135559

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:720]: https://github.com/memgraph/memgraph/pull/566#discussion_r1016135559

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py]: This was rewritten, it is not repetitive anymore.

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:728]: I need to balance this out bcs of that long-running expansion test, I will do this on the fly, today/tomorrow.

File: tests/mgbench/datasets.py
Comment (antejavor) [reply tests/mgbench/datasets.py:50]: I synced with @jbajic about this. Thanks for the tip.

**Comment by antejavor** [comment]
@42jeremy Regarding your comment 😀, we can hop on a call tomorrow morning to sync about these changes.

File: tests/mgbench/benchmark.py
Comment (42jeremy) [inline tests/mgbench/benchmark.py:142]: If I understand correctly the first sentence, how the mixed-workload is interpreted depends on the other arguments (= single configuration VS defined load). If I am correct, would it make sense to provide a command/example for each? (feel free to disagree and resolve this comment yourself)

I am providing the following as an example, please correct and adapt if it does not make sense:

`./benchmark dataset/small --mixed-workload 1000 20 70 10 0:` run on dataset `small` from folder `dataset`, total of 1000 queries 20% write etc..
`./benchmark dataset/small/read_specific --mixed-workload 1000 25 0 20 10 45`:  run on dataset `small` from folder `dataset`, total of 1000 queries, 45% will be from read_specific, 25% write, 20% update, 10% analytical

File: tests/mgbench/datasets.py
Comment (42jeremy) [inline tests/mgbench/datasets.py:134]: Adresses of medium and large should change as well

Comment (42jeremy) [reply tests/mgbench/datasets.py:134]: https://s3.eu-west-1.amazonaws.com/deps.memgraph.io/dataset/pokec/pokec_small.setup.cypher
https://s3.eu-west-1.amazonaws.com/deps.memgraph.io/dataset/pokec/pokec_medium.setup.cypher
https://s3.eu-west-1.amazonaws.com/deps.memgraph.io/dataset/pokec/pokec_large.setup.cypher.gz

(from: https://github.com/memgraph/memgraph/blob/5c0e41ed44c283d172ecb9106bbc5809e8cb2906/tests/mgbench/datasets.py#L106-L110 )

Comment (42jeremy) [reply tests/mgbench/datasets.py:134]: (maybe that's something we'll tackle when merging your PR to project-pineapple, so you can ignore this comment)

File: tests/mgbench/benchmark.py
Comment (antejavor) [reply tests/mgbench/benchmark.py:142]: I will add a bit more explanation in the next commit, and this  PR will have to follow up polish PR I will make this thing more straightforward.

**Comment by antejavor** [comment]
Things that can be added to follow-up PR: 

- Polish mixed workload flags
- Add detailed logging
- Add more detailed resource tracking
- Polish `benchmark.py` script structure
- More vendors (future)
- Polish C++ wrapper (future)

File: tests/mgbench/benchmark.py
Comment (gitbuda) [reply tests/mgbench/benchmark.py:32]: Both have to run on CI

Comment (antejavor) [reply tests/mgbench/benchmark.py:32]: This is fixed now, both non-auth and auth options are running on CI.

**Review by gitbuda — APPROVED** [review]

**Pull request merged**
